### PR TITLE
Aos_2: Change Make_x_monotone_2 return type from legacy CGAL::Object to boost::variant

### PR DESCRIPTION
**Pull request by efifogel**
## Summary of Changes

Change the type of the result of the Make_x_monotone_2 functor operator. More precisely, the old type was an output iterator that dereferenced CGAL::Object(). The new type is an output-iterator that dereferences variant of the possible types of results. (Both wrap either Point_2 and X_monotone_2 objects.
Also fixed some left overs of the previous change of the result type of the Intersect_2 functor from legacy CGAL::Object to boost::variant.

## Release Management

* Affected package(s): Arrangement_on_surface_2
* Issue(s) solved (if any): fix #0000, fix #0000,...
* Feature/Small Feature (if any): No (backward compatible)
* Link to compiled documentation (obligatory for small feature) [*wrong link name to be changed*](httpssss://wrong_URL_to_be_changed/Manual/Pkg)
* License and copyright ownership: TAU

I named the repo Aos_2-remove_object-efif cause I thought that I would eliminate all uses of CGAL::Object, but now, I've decided to split it into 2 steps, so there will be another round that will take care of the remaining uses in insert(), arr.zone(), and vertical_decomposition().

### CODE REVIEW

**Comment by efifogel** [comment]
I believe that I've fixed an issue that caused clang to fail.
Anyway, the code is ready for testing and integration.

File: Arrangement_on_surface_2/include/CGAL/Arr_curve_data_traits_2.h [hunk 1]
```diff
@@ -27,7 +27,6 @@
 #include <boost/utility/enable_if.hpp>
 #include <boost/mpl/has_xxx.hpp>
 
-#include <CGAL/Object.h>
 #include <CGAL/tags.h>
 #include <CGAL/Arr_tags.h>
 #include <CGAL/Arr_geometry_traits/Curve_data_aux.h>
```
Comment (lrineau) [inline Arrangement_on_surface_2/include/CGAL/Arr_curve_data_traits_2.h:32]: This header requires `<CGAL/assertions.h>`. See the Travis error:
https://travis-ci.com/github/CGAL/cgal/jobs/357577087#L7296

That header was probably pulled by `<CGAL/Object.h>`.

**Comment by efifogel** [comment]
Fixed (added the missing include statement).
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by lrineau** [comment]
@efifogel Please fix the conflicts with `master` (with a merge), so that the PR can be tested.

**Comment by efifogel** [comment]
Done
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

File: Installation/CHANGES.md [hunk 2]
```diff
@@ -110,6 +116,13 @@ Release date: September 2020
      As there exists an implicit conversion from `boost::variant` to `CGAL::Object`, the
      new code is backward compatible. However, it is recommended that all calls
      to the intersection functions are fixed to use the new return type.
+ -   Changed `Make_x_monotone_2::operator()(`) return type from legacy [`CGAL::Object`](https://doc.cgal.org/5.1/STL_Extension/classCGAL_1_1Object.html)
+     to modern `boost::variant` in all traits concepts and models. As exaplained
+     above, the code is backward compatible. However, it is recommended that all
+     calls to `Make_x_monotone_2::operator()()` are fixed to use the new return type.
+ -   Changed `decompose()` interface to use modern `boost::variant` instead of legacy [`CGAL::Object`](https://doc.cgal.org/5.1/STL_Extension/classCGAL_1_1Object.html)
+     As exaplained above, the code is backward compatible. However, it is recommended that all
+     calls to `decompose()` are fixed to use the new interface.
```
Comment (lrineau) [inline Installation/CHANGES.md:125]: This part of the patch is in the section about CGAL-5.1. It should be in 5.2.

**Comment by efifogel** [comment]
Sorry, my silly mistake makes you work harder...
Anyhow, I've fixed it.
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by lrineau** [comment]
Thanks Efi. This PR will be tested tonight, if the Travis checks run fine.

**Comment by maxGimeno** [comment]
There are warnings : https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.2-Ic-10/Arrangement_on_surface_2/TestReport_gimeno_Fedora-rawhide.gz

**Comment by efifogel** [comment]
Fixed
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by maxGimeno** [comment]
Warnings :
 - https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.2-Ic-11/Arrangement_on_surface_2_Examples/TestReport_lrineau_Fedora-Release.gz
 - https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.2-Ic-11/Boolean_set_operations_2_Examples/TestReport_afabri_x64_Cygwin-Windows10_MSVC2017-Release-64bits.gz
 - https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.2-Ic-11/CGAL_ipelets_Demo/TestReport_lrineau_Debian-stable-Release.gz

Error: 
 - https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.2-Ic-11/Arrangement_on_surface_2_Demo/TestReport_afabri_x64_Cygwin-Windows10_MSVC2017-Debug-64bits.gz

**Comment by efifogel** [comment]
I've fixed the warnings.
The error probably also happens in master, as it has nothing to do with
this PR.
It will have to be fixed in a different PR.
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by sloriot** [comment]
The error is probably from #4942

**Comment by efifogel** [comment]
The error message is cryptic, and I cannot reproduce it, but it's a
compilation error, and I think it should be easy to fix (had I known the
source of it...)
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by lrineau** [comment]
> The error message is cryptic, and I cannot reproduce it, but it's a compilation error, and I think it should be easy to fix (had I known the source of it...) 

Well... Let's see:
```
C:\CGAL\test\CGAL-5.2-Ic-11\include\CGAL/Curved_kernel_via_analysis_2/gfx/Curve_renderer_2.h(1753): error C2760: syntax error: unexpected token 'identifier', expected ')'
```

That line 1753 is the first line of the file where a variable named `_` is used. My guess is that the name `_` is the reason of the compilation error. Maybe a header from a dependency defines `_` as a macro.

**Comment by efifogel** [comment]
Probably something like this. In any case, nothing to do with this PR
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by maxGimeno** [comment]
There is a conflict in CHANGES.md

**Comment by efifogel** [comment]
It seems that I'm chasing my tail.
Anyhow, I've merged with master and resolved the conflict yet again.
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by maxGimeno** [comment]
Thanks !

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.2-Ic-18.shtml

**Pull request merged**